### ISSUE TITLE
Supporting GNOME Web / Epiphany web browser

### ISSUE THREAD

**Issue by Nico7as** [nextcloud/server#40793]
I got a warning message about my browser not being supported (Epiphany a.k.a. Gnome Web) the last time logged in my Nextcloud account.

I reached out to the Epiphany developers and after some researches, they told me that the issue came from NextCloud.
Might be "user agent header" related, or It might come from an API incompatibility with WebKitGTK.

Since NextCloud is well integrated to Gnome, I naively thought that everything would be 100% compatible between both plateforms.

I understand that Epiphany is quite confidential, being exclusive to Linux, where Firefox and Chromium/Chrome are way more popular.
Still, it's a polished and well integrated browser, part of the Gnome DE, which is a very popular DE among mainline distributions. And it allows to "install" websites as webapps. Very convinient for NextCloud's Deck, for exemple.

I wish you could make NC Epiphany friendly :-)

Cheers

**Comment by mcatanzaro** [comment]
OK, problem here is all the information required to fix this issue is in your [Epiphany issue report](https://gitlab.gnome.org/GNOME/epiphany/-/issues/2195), but you did not link to it, greatly reducing the chances of this being fixed. So, following that link is important. ;)

Some notes:

 * WebKitGTK advertises as Safari 16.4, which is apparently supported by Nextcloud. The standard user agent header is `Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.4 Safari/605.1.15` and Epiphany does not modify it.
 * Since unsupported browser warnings are extremely antagonistic to browser developers and hostile to users, I add quirks to WebKit whenever required to suppress them. Websites that display these warnings to WebKit users simply lose the right to receive an accurate user agent header from WebKit.
 * The ideal solution is for Nextcloud to just not do this. That would be very nice. :)
 * I tested by visiting https://acloud2.zaclys.com/index.php/login and I do not see the warning myself, so I can't help develop a quirk for this as I usually would. But I'm still willing to add one if this doesn't get fixed in Nextcloud and you can find one that works.

**Comment by mcatanzaro** [comment]
Oh and I eventually figured out that browesrslist is an external project https://github.com/browserslist/browserslist-useragent-regexp but haven't looked closer than that.

**Comment by joshtrichards** [comment]
Browser compatibility checking (at least any centralized) seems defined here:

https://github.com/nextcloud-libraries/browserslist-config/blob/master/browserlist.config.js

You can see what this translates to by running that filter through the library's checker here:

https://browsersl.ist/#q=%3E0.25%25%2C+not+op_mini+all%2C+not+dead%2C+Firefox+ESR

(with some variance depending on NC version and build time)

Interestingly, I don't see Safari 16.4 on the list...

**Comment by Nico7as** [comment]
@mcatanzaro 
Thanks for adding the link, and relevant info !

I get the warning message after the logging in.

And I forgot to give my versions:

- Fedora 38
- Nextcloud 26 (or 25, I'm not sure)
- GNOME Web 44.6

**Comment by mcatanzaro** [comment]
> Interestingly, I don't see Safari 16.4 on the list...

Well that's surely the problem. The Nextcloud UI _says_ that 16.4 is supported, but its caniuse usage data is lower than 0.25% so it's excluded from your browserlist filter. Surely Nextcloud's UI shouldn't be inconsistent with the filter. If you change the filter to `>0.25%, Safari >= 16.4, not op_mini all, not dead, Firefox ESR` then I bet it would work.

But I assume many other websites are likely using browserlist too. I never imagined that these unsupported browser warnings would be based on caniuse.com data. It's probably time for me to bump the fake Safari version in WebKitGTK's user agent header -- I normally do this once or twice per year -- but I would bump it to the latest version 17.0, which is also apparently < 0.25% and therefore also won't be matched by your query. The `defaults` query includes `last 2 versions` to address this, but it's missing from your config. So I would add that too.

Unfortunately browserlist is counting each Safari minor version rather than only counting the major versions, so the last two versions are actually 17.0 and 16.6. It seems `last 2 Safari major versions` would avoid this, but that's not part of `defaults` so I can't expect websites to use it. Sigh.

I will bump our fake Safari version to 17.0 now. I wonder how many websites would break if I set it to something higher than the latest version....

**Comment by joshtrichards** [comment]
It seems we also have `allowHigherVersions` toggled on:

https://github.com/nextcloud/server/blob/71cd97c37f9879c3c8ee06e8a56f8995b0195f45/core/src/services/BrowsersListService.js#L29

This *should* cover Safari v16.4 just fine (and, in turn, Epiphany).

Anyhow, I installed Epiphany on my Debian 12 workstation and reproduced the behavior only in the latest major version of NC (v27), but neither of the still supported major versions (v25, v26) consider Epiphany unsupported.

So I think we're running into an inadvertent regression, possibly from bumping up a dependency (like browserslist-useragent).

And the warning here about using `allowHigherVersions`:

https://github.com/browserslist/browserslist-useragent#when-querying-for-modern-browsers

...makes me wonder if we should switch to `unreleased versions` + `ignoreMinor` instead.

**Comment by Nico7as** [comment]
I've asked on the forum. I'll give you the answer as soon as I can.

I'm not running NC my self, but subscribing to it as a service.
I assumed the server weren't upgraded to NC27 yet, cause it seems to recent.
The Zaclys company is small, they upgrade the shared servers when they can.

**Comment by joshtrichards** [comment]
Is it their shared server at https://ncloud.zaclys.com/ or a dedicated one?

If the former, it looks like it's v24.0.11 (Enterprise Edition).

If dedicated, you can visit your https://domain.tld/status.php on your server to probably find out.

**Comment by Nico7as** [comment]
no, mine is 
https://acloud2.zaclys.com

"nclouds" are free, and basics
"aclouds" are for paid subscribers (a = abonné in french = subscribers)
they're updated more often

**Comment by joshtrichards** [comment]
Cool. You're running v27.1.1. :-)

Okay good. That matches my testing. So this is some sort of regression it seems.

**Comment by mcatanzaro** [comment]
For completeness: [WebKit bug to bump Safari version](https://bugs.webkit.org/show_bug.cgi?id=262782)

**Comment by Nico7as** [comment]
@joshtrichards Yes, and now it know how to find it by myself :-). Didn't see the arrow on the left, to get more details...

thanks again you both @mcatanzaro & @joshtrichards  for spending time on this !!

Makes me want to give more user feedbacks to the softwares I use !

**Comment by joshtrichards** [comment]
I no longer see this behavior in any supported major releases (i.e. >= 28).

**Comment by mcatanzaro** [comment]
Sure, but it's just going to happen again (and again, and again) if you don't change your policy for user agent version checking....

**Comment by joshtrichards** [comment]
It's a trade-off: people complain when things aren't working because they don't check the requirements listed in our docs or they get a warning when they're running what (is usually) an outdated browser. 

Keep in mind that this is a warning that is displayed to end-users of Nextcloud instances; not just admins and other people that understand what is going on.

There's no ideal solution here I guess. :shrug:

**Comment by joshtrichards** [comment]
Actually: https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/config_sample_php_parameters.html#no-unsupported-browser-warning 

:-)

**Comment by mcatanzaro** [comment]
What I'll probably do is change WebKitGTK to set a very high Safari version like Version/99.9 to fix Nextcloud and hope that it fixes more websites than it breaks. We'll see how many websites decide that high versions are unsupported....

Any attempt to decide whether the browser is supported based on the user agent header is misguided because the Safari versions have to be fake, but I can't stop you from trying....

**Comment by nextcloud-command** [comment]
This issue has been automatically marked as stale because it has not had recent activity and seems to be missing some essential information. It will be closed if no further activity occurs. Thank you for your contributions.
